Docs: Document caching mechanisms (#52514)
vercel/next.js: docs/02-app/01-building-your-application/02-data-fetching/01-fetching-caching-and-revalidating.mdx

export default async function Page() {
  const data = await getData()

  return <main></main>
}
```

> **Good to know**:
>
> Next.js provides helpful functions you may need when fetching data in Server Components such as [`cookies`](/docs/app/api-reference/functions/cookies) and [`headers`](/docs/app/api-reference/functions/headers). These will cause the route to be dynamically rendered as they rely on request time information.
> In Route handlers, `fetch` requests are not memoized as Route Handlers are not part of the React component tree.
> To use `async`/`await` in a Server Component with TypeScript, you'll need to use TypeScript `5.1.3` or higher and `@types/react` `18.2.8` or higher.

### Caching Data

Caching stores data so it doesn't need to be re-fetched from your data source on every request.

By default, Next.js automatically caches the returned values of `fetch` in the [Data Cache](/docs/app/building-your-application/caching#data-cache) on the server. This means that the data can be fetched at build time or request time, cached, and reused on each data request.

```js
// 'force-cache' is the default, and can be omitted
fetch('https://...', { cache: 'force-cache' })
```

> `fetch` requests that use the `POST` method are also automatically cached in the Data Cache – unless it's used in `POST` [Route Handler](/docs/app/building-your-application/routing/route-handlers) or comes after `headers`/`cookies`.

Reviewer: Perhaps `after` could be expanded a bit to clarify that it's about being after in terms of lines of code / execution order.

Author: Good point, thank you!

export default async function Page() {
  const data = await getData()

  return <main></main>
}
```

> **Good to know**:
>
> Next.js provides helpful functions you may need when fetching data in Server Components such as [`cookies`](/docs/app/api-reference/functions/cookies) and [`headers`](/docs/app/api-reference/functions/headers). These will cause the route to be dynamically rendered as they rely on request time information.
> In Route handlers, `fetch` requests are not memoized as Route Handlers are not part of the React component tree.
> To use `async`/`await` in a Server Component with TypeScript, you'll need to use TypeScript `5.1.3` or higher and `@types/react` `18.2.8` or higher.

### Caching Data

Caching stores data so it doesn't need to be re-fetched from your data source on every request.

By default, Next.js automatically caches the returned values of `fetch` in the [Data Cache](/docs/app/building-your-application/caching#data-cache) on the server. This means that the data can be fetched at build time or request time, cached, and reused on each data request.

```js
// 'force-cache' is the default, and can be omitted
fetch('https://...', { cache: 'force-cache' })
```

> `fetch` requests that use the `POST` method are also automatically cached in the Data Cache – unless it's used in `POST` [Route Handler](/docs/app/building-your-application/routing/route-handlers) or comes after `headers`/`cookies`.

Reviewer 2: At least current implementation, it is true. 

Also, please refer to https://github.com/vercel/next.js/pull/52669#issuecomment-1638295085 I'm not sure this my comment is right because I have not recieved any response from the Vercel staff...

> `fetch` requests that use the `POST` method are also automatically cached in the Data Cache – unless it's used in `POST` [Route Handler](/docs/app/building-your-application/routing/route-handlers) or comes after `headers`/`cookies`.

Author: I've grouped every case and asked for clarity on Slack. I think it's a lot more nuanced, and it'll be good to dig into the commerce use case.

Cached data can be revalidated in two ways:

- **Time-based revalidation**: Automatically revalidate data after a certain amount of time has passed. This is useful for data that changes infrequently and freshness is not as critical.
- **On-demand revalidation**: Manually revalidate data based on an event (e.g. form submission). On-demand revalidation can use a tag-based or path-based approach to revalidate groups of data at once. This is useful when you want to ensure the latest data is shown as soon as possible (e.g. when content from your headless CMS is updated).

#### Time-based Revalidation

To revalidate data at a timed interval, you can use the `next.revalidate` option of `fetch` to set the cache lifetime of a resource (in seconds).

```js
fetch('https://...', { next: { revalidate: 3600 } })
```

Alternatively, to revalidate all `fetch` requests in a route segment, you can use the [Segment Config Options](/docs/app/api-reference/file-conventions/route-segment-config).

```jsx filename="layout.js / page.js"

> export const revalidate = 3600 // revalidate every hour

Author: Yes, thank you, let's add 'at most'

  const data = await res.json()
  // ...
}
```

```jsx filename="app/page.js" switcher
export default async function Page() {
  const res = await fetch('https://...', { next: { tags: ['collection'] } })
  const data = await res.json()
  // ...
}
```

The data can then be revalidated on-demand by calling `revalidateTag` in a [Route Handler](/docs/app/building-your-application/routing/route-handlers) or [Server Action](/docs/app/building-your-application/data-fetching/server-actions).

```ts filename="app/api/revalidate/route.ts" switcher
import { NextRequest, NextResponse } from 'next/server'
import { revalidateTag } from 'next/cache'

export async function GET(request: NextRequest) {
  const tag = request.nextUrl.searchParams.get('tag')
  revalidateTag(tag)
  return NextResponse.json({ revalidated: true, now: Date.now() })
}
```

```js filename="app/api/revalidate/route.js" switcher
import { NextResponse } from 'next/server'
import { revalidateTag } from 'next/cache'

export async function GET(request) {
  const tag = request.nextUrl.searchParams.get('tag')
  revalidateTag(tag)
  return NextResponse.json({ revalidated: true, now: Date.now() })
}
```

> Alternatively, you can use [`revalidatePath`](/docs/app/api-reference/functions/revalidatePath) in a [Route Handler](/docs/app/building-your-application/routing/route-handlers) or [Server Action](/docs/app/building-your-application/data-fetching/server-actions) to revalidate all data associated with a path.

Author: Will add it!

Learn more about [on-demand revalidation](/docs/app/building-your-application/caching#on-demand-revalidation).

#### Error handling and revalidation

If an error is thrown while attempting to revalidate data, the last successfully generated data will continue to be served from the cache. On the next subsequent request, Next.js will retry revalidating the data.

### Opting out of Data Caching

#### Individiual `fetch` Requests

> To opt individual `fetch` requests out of caching, you can set the `cache` option in `fetch` to `'no-store'`. This will fetch data dynamically, on every request.

Author: I believe the first one is correct, as "out" needs to be followed by "of". Though we could probably improve this sentence grammatically with an alternative: "To opt out of caching for individual `fetch` requests".

```js
fetch('https://...', { cache: 'no-store' })
```

View all the available `cache` options in the [`fetch` API reference](/docs/app/api-reference/functions/fetch).

> Requests are also **not** cached if:
>

> > - The `fetch` request uses the `POST` method, the `Authorization` header, or `cookie` headers.

Reviewer 2: I don't know if this is intentional or not, but in fact the `POST` method is cached even if `Authorization` or `Cookie` headers are present, except for dynamic rendering (after `headers()`/`cookies()`, `cache: "no-store"`, `revalidate: 0`, etc.).
Note line 221 below:
https://github.com/vercel/next.js/blob/afa9f965157ce60b5d65d1b3d8981e57b06d8e9b/packages/next/src/server/lib/patch-fetch.ts#L216-L221
Therefore, I think this line should be removed:
```suggestion
```

> > - The `fetch` request uses the `POST` method, the `Authorization` header, or `cookie` headers.

Author: Thank you @koichik!

> - The `fetch` request is inside a Router Handler that uses the `POST` method.
> - `revalidate: 0` is configured on individual `fetch` requests
> - The `fetchCache` route segment option is configured to skip cache by default.

#### Multiple `fetch` Requests

If you have multiple `fetch` requests in a route segment (e.g. a Layout or Page), you can configure the caching behavior of all data requests in the segment using the [Segment Config Options](/docs/app/api-reference/file-conventions/route-segment-config).

For example, using `const dynamic = 'force-dynamic` will cause all data to be fetched at request time, and the segment to be rendered dynamically.

> ```ts

Author: Will do!

// Add
export const dynamic = 'auto'
```

There's an extensive list of Segment Config options, giving you fine-grained control of static and dynamic behavior of a route segment. See the [API reference](/docs/app/api-reference/file-conventions/route-segment-config) for more.

> ## Fetching data with third-party libraries

Author: I hear you, if someone is already using axios, react `cache` and `unstable_cache` might be a way for them to quickly migrate without moving to `fetch`. Leaving it out makes it clear that it's not a recommendation.

}
```

Alternatively, you can use [`revalidatePath`](/docs/app/api-reference/functions/revalidatePath) in a [Route Handler](/docs/app/building-your-application/routing/route-handlers) or [Server Action](/docs/app/building-your-application/data-fetching/server-actions) to revalidate all data associated with a path.

Learn more about [on-demand revalidation](/docs/app/building-your-application/caching#on-demand-revalidation).

#### Error handling and revalidation

If an error is thrown while attempting to revalidate data, the last successfully generated data will continue to be served from the cache. On the next subsequent request, Next.js will retry revalidating the data.

### Opting out of Data Caching

`fetch` requests are **not** cached if:

- The `cache: 'no-store` is added to `fetch` requests.
- The `revalidate: 0` option is added to individual `fetch` requests.
- The `fetch` request is inside a Router Handler that uses the `POST` method.
- The `fetch` request comes after the usage of `headers` or `cookies`.
- The `const dynamic = 'force-dynamic'` route segment option is used.
- The `fetchCache` route segment option is configured to skip cache by default.

> - The `fetch` request uses `Authorization` or `Cookie` headers and there's an uncached request above it in the component tree.

Author: Need to verify if these cases are correct

---
title: Data Fetching, Caching, and Revalidating
nav_title: Fetching, Caching, and Revalidating
description: Learn how to fetch, cache, and revalidate data in your Next.js application.
---

Data fetching is a core part of any application. This page goes through how you can fetch, cache, and revalidate data in React and Next.js.

There are three main ways you can fetch data:

1. On the server, with the `fetch` API.
2. On the server, with third-party libraries.
3. On the client, with third-party libraries.

## Fetching Data on the Server with `fetch`

> Next.js extends the native [`fetch` Web API](https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API) to allow you to configure the [caching](#caching-data) and [revalidating](#revalidating-data) behavior for each fetch request on the server. React extends `fetch` to automatically [memoize](/docs/app/building-your-application/data-fetching/patterns#fetching-data-where-its-needed) fetch requests while rendering a React Component tree.

Reviewer 3: ```suggestion
Next.js extends the native [`fetch` Web API](https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API) to allow you to configure the [caching](#caching-data) and [revalidating](#revalidating-data) behavior for each fetch request on the server. React extends `fetch` to automatically [memoize](/docs/app/building-your-application/data-fetching/patterns#fetching-data-where-its-needed) fetch requests while rendering a React component tree.
```

> You can use `fetch` with [`async`/`await` in Server Components](https://github.com/acdlite/rfcs/blob/first-class-promises/text/0000-first-class-support-for-promises), in [Route Handlers](), and in [Server Actions]().

Reviewer 3: ```suggestion
You can use `fetch` with [`async`/`await` in Server Components](https://github.com/acdlite/rfcs/blob/first-class-promises/text/0000-first-class-support-for-promises), in [Route Handlers](/docs/app/building-your-application/routing/router-handlers), and in [Server Actions](/docs/app/building-your-application/data-fetching/server-actions).
```

    // This will activate the closest `error.js` Error Boundary
    throw new Error('Failed to fetch data')
  }

  return res.json()
}

export default async function Page() {
  const data = await getData()

  return <main></main>
}
```

```jsx filename="app/page.js" switcher
async function getData() {
  const res = await fetch('https://api.example.com/...')
  // The return value is *not* serialized
  // You can return Date, Map, Set, etc.

  if (!res.ok) {
    // This will activate the closest `error.js` Error Boundary
    throw new Error('Failed to fetch data')
  }

  return res.json()
}

export default async function Page() {
  const data = await getData()

  return <main></main>
}
```

> **Good to know**:
>
> To use `async`/`await` in a Server Component with TypeScript, you'll need to use TypeScript `5.1.3` or higher and `@types/react` `18.2.8` or higher.
> Next.js provides helpful functions you may need when fetching data in Server Components such as [`cookies`](/docs/app/api-reference/functions/cookies) and [`headers`](/docs/app/api-reference/functions/headers). These will cause the route to be dynamically rendered as they rely on request time information.

### Caching Data

> Caching is the process of storing data on the server (e.g. an [Edge Network](https://vercel.com/docs/concepts/edge-network/overview)) so it doesn't need to be re-fetched from your data source each time `fetch` is called.

Reviewer 3: ```suggestion
Caching stores data so it doesn't need to be re-fetched from your data source each time `fetch` is called.
```

Thinking about removing this because it might make it seem like you need an edge network for caching.

By default, Next.js automatically caches the returned values of `fetch` in the [Data Cache](/docs/app/building-your-application/caching#data-cache) on the server. This means that the data can be fetched at build time or request time, cached, and reused on each data request.

```js
// 'force-cache' is the default, and can be omitted
fetch('https://...', { cache: 'force-cache' })
```

> `fetch` requests that use the `POST` method are also automatically cached in the Data Cache – unless it's used in `POST` [Route Handler](/docs/app/building-your-application/routing/route-handlers) or come after `headers`/`cookies`.

Reviewer 3: ```suggestion
`fetch` requests that use the `POST` method are also automatically cached in the Data Cache – unless it's used in `POST` [Route Handler](/docs/app/building-your-application/routing/route-handlers) or comes after `headers`/`cookies`.
```

> **How does the Data Cache work?**
>

> > The Data Cache is a persistent [HTTP cache](https://developer.mozilla.org/en-US/docs/Web/HTTP/Caching). Depending on your platform, the cache can scale automatically and be shared across multiple regions.

Reviewer 3: ```suggestion
> The Data Cache is a persistent [HTTP cache](https://developer.mozilla.org/en-US/docs/Web/HTTP/Caching). Depending on your platform, the cache can scale automatically and be [shared across multiple regions](https://vercel.com/docs/infrastructure/data-cache).
```

>
> Next.js extends the [options object](https://developer.mozilla.org/en-US/docs/Web/API/fetch#:~:text=preflight%20requests.-,cache,-A%20string%20indicating) of the `fetch` function to allow each request on the server to set its own persistent caching behavior on the server.
> During rendering, when Next.js comes across a fetch, it will check the cache to see if the data is already available. If it is, it will return the cached data. If not, it will fetch and store data for future requests.
>
> Learn more about the [Data Cache](/docs/app/building-your-application/caching#data-cache).

### Revalidating Data

Revalidation is the process of purging the Data Cache and re-fetching the latest data. This is useful when your data changes and you want to ensure you show the latest information.

Cached data can be revalidated in two ways:

> - **Time-based revalidation**: Automatically revalidate data after a certain amount of time has passed. This is useful for data that changes infrequently and freshness is not as critical. Time-based revalidation can be configured with the [Route Segment Config options](/docs/app/api-reference/file-conventions/route-segment-config) or per individual `fetch` request.

Reviewer 3: ```suggestion
- **Time-based revalidation**: Automatically revalidate data after a certain amount of time has passed. This is useful for data that changes infrequently and freshness is not as critical. Time-based revalidation can be configured with the [Route Segment Config options](/docs/app/api-reference/file-conventions/route-segment-config) or on individual `fetch` requests.
```

> - **On-demand revalidation**: Manually revalidate data based on an event (e.g. form submission). On-demand revalidation can use a tag-based or path-based approach to revalidate groups of data at once. This is useful when you want to ensure the latest data is shown as soon as possible, or content from your headless CMS is updated. On-demand revalidation can be triggered with [`revalidateTag`](/docs/app/api-reference/functions/revalidateTag) or [`revalidatePath`](/docs/app/api-reference/functions/revalidatePath).

Reviewer 3: ```suggestion
- **On-demand revalidation**: Manually revalidate data based on an event (e.g. form submission). On-demand revalidation can use a tag-based or path-based approach to revalidate groups of data at once. This is useful when you want to ensure the latest data is shown as soon as possible (e.g. when content from your headless CMS is updated). On-demand revalidation can be triggered with [`revalidateTag`](/docs/app/api-reference/functions/revalidateTag) or [`revalidatePath`](/docs/app/api-reference/functions/revalidatePath).
```

>
> This is similar to [**stale-while-revalidate**](https://web.dev/stale-while-revalidate/) behavior.

#### On-demand Revalidation

Data can be revalidated on-demand by path ([`revalidatePath`](/docs/app/api-reference/functions/revalidatePath)) or by cache tag ([`revalidateTag`](/docs/app/api-reference/functions/revalidateTag)).

Next.js has a cache tagging system for invalidating `fetch` requests across routes.

1. When using `fetch`, you have the option to tag cache entries with one or more tags.
2. Then, you can call `revalidateTag` to revalidate all entries associated with that tag.

For example, the following `fetch` requests adds the cache tag `collection`:

```tsx filename="app/page.tsx" switcher
export default async function Page() {
  const res = await fetch('https://...', { next: { tags: ['collection'] } })
  const data = await res.json()
  // ...
}
```

```jsx filename="app/page.js" switcher
export default async function Page() {
  const res = await fetch('https://...', { next: { tags: ['collection'] } })
  const data = await res.json()
  // ...
}
```

> The data can then be revalidated on-demand by calling `revalidateTag` in a [Route Handler](/docs/app/building-your-application/routing/route-handlers) or [Server Action]().

Reviewer 3: ```suggestion
The data can then be revalidated on-demand by calling `revalidateTag` in a [Route Handler](/docs/app/building-your-application/routing/route-handlers) or [Server Action](/docs/app/building-your-application/data-fetching/server-actions).
```

```ts filename="app/api/revalidate/route.ts" switcher
import { NextRequest, NextResponse } from 'next/server'
import { revalidateTag } from 'next/cache'

export async function GET(request: NextRequest) {
  const tag = request.nextUrl.searchParams.get('tag')
  revalidateTag(tag)
  return NextResponse.json({ revalidated: true, now: Date.now() })
}
```

```js filename="app/api/revalidate/route.js" switcher
import { NextResponse } from 'next/server'
import { revalidateTag } from 'next/cache'

export async function GET(request) {
  const tag = request.nextUrl.searchParams.get('tag')
  revalidateTag(tag)
  return NextResponse.json({ revalidated: true, now: Date.now() })
}
```

> Alternatively, you can use [`revalidatePath`](/docs/app/api-reference/functions/revalidatePath) in a [Route Handler]() or [Server Action]() to revalidate all data associated with a path.

Reviewer 3: ```suggestion
Alternatively, you can use [`revalidatePath`](/docs/app/api-reference/functions/revalidatePath) in a [Route Handler](/docs/app/building-your-application/routing/route-handlers) or [Server Action](/docs/app/building-your-application/data-fetching/server-actions) to revalidate all data associated with a path.
```

#### Error handling and revalidation

If an error is thrown while attempting to revalidate data, the last successfully generated data will continue to be served from the cache. On the next subsequent request, Next.js will retry revalidating the data.

### Opting out of Data Caching

#### Individiual `fetch` Requests

To opt individual `fetch` requests out of caching, you can set the `cache` option in `fetch` to `'no-store'`. This will fetch data dynamically, on every request.

```js
fetch('https://...', { cache: 'no-store' })
```

View all the available `cache` options in the [`fetch` API reference](/docs/app/api-reference/functions/fetch).

> Requests are also **not** cached if:
>
> - The `fetch` request uses the `POST` method, the `Authorization` header, or `cookie` headers.
> - The `fetch` request is inside a Router Handler that uses the `POST` method.
> - `revalidate: 0` is configured on individual `fetch` requests
> - The `fetchCache` route segment option is configured to skip cache by default.

#### Multiple `fetch` Requests

> If you have multiple `fetch` requests in a route segment (e.g. a layout or page), you can configure the caching behavior of all data requests in the segment using the [Segment Config Options]().

Reviewer 3: ```suggestion
If you have multiple `fetch` requests in a route segment (e.g. a layout or page), you can configure the caching behavior of all data requests in the segment using the [Segment Config Options](/docs/app/api-reference/file-conventions/route-segment-config).
```

For example, using `const dynamic = 'force-dynamic` will cause all data to be fetched at request time, and the segment to be rendered dynamically.

```ts
// Add
export const dynamic = 'auto'
```

There's an extensive list of Segment Config options, giving you fine-grained control of static and dynamic behavior of a route segment. See the [API reference](/docs/app/api-reference/file-conventions/route-segment-config) for more.

## Fetching data with third-party libraries

> In cases where you're using a third-party library that doesn't support or expose `fetch` (for example, a database, CMS, or ORM client), you can configure the caching and revalidating behavior of those requests using the [Route Segment Config Option]() and React's `cache` function.

Reviewer 3: ```suggestion
In cases where you're using a third-party library that doesn't support or expose `fetch` (for example, a database, CMS, or ORM client), you can configure the caching and revalidating behavior of those requests using the [Route Segment Config Option](/docs/app/api-reference/file-conventions/route-segment-config) and React's `cache` function.
```

Whether the data is cached or not will depend on whether the route segment is [statically or dynamically rendered](). If the segment is static (default), the output of the request will be cached and revalidated as part of the route segment. If the segment is dynamic, the output of the request will _not_ be cached and will be re-fetched on every request when the segment is rendered.

> **Good to know:**
>

> > In the future, Next.js will have a API for configuring the caching and revalidating behavior of individual third-party requests. Learn more: [`unstable_cache`]()

Reviewer 3: Maybe drop the learn more?

### Example

In the example below:

> - The `revalidate` option is set to `3600`, meaning the data will be cached and revalidated every hour.

Reviewer 3: ```suggestion
- The `revalidate` option is set to `3600`, meaning the data will be cached and revalidated at most every hour.
```

> - The React `cache` function is used to [memoize]() data requests.

Reviewer 3: TODO

```ts filename="utils/getUser.ts" switcher
import { cache } from 'react'

> export const revalidate = 3600 // revalidate the data every hour
> 
> export const getUser = cache(async (id: string) => {
>   const user = await db.user.findUnique({ id })
>   return user
> })
> ```
> 
> ```js filename="utils/getUser.js" switcher
> import { cache } from 'react'
> 
> export const revalidate = 3600 // revalidate the data every hour

Reviewer 3: ```suggestion
export const revalidate = 3600 // revalidate the data at most every hour

export const getUser = cache(async (id: string) => {
  const user = await db.user.findUnique({ id })
  return user
})
```

```js filename="utils/getUser.js" switcher
import { cache } from 'react'

export const revalidate = 3600 // revalidate the data at most every hour
```
